src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Build/SolutionBuildManager.cs

```diff
@@ -168,4 +173,88 @@ public void StartUpdateSpecificProjectConfigurations(IVsHierarchy[] hierarchy, u
             dwFlags: dwFlags,
             fSuppressUI: 0));
     }
+
+    public async ValueTask<bool> BuildProjectAndWaitForCompletionAsync(
+        IVsHierarchy projectHierarchy,
+        CancellationToken cancellationToken = default)
+    {
+        return await _semaphore.ExecuteAsync(async () => await BuildProjectCoreAsync(), cancellationToken);
+
+        async Task<bool> BuildProjectCoreAsync()
+        {
+            await InitializeAsync();
+            Assumes.NotNull(_vsSolutionBuildManager2);
+            await JoinableFactory.SwitchToMainThreadAsync(cancellationToken);
+            using var solutionBuildCompleteListener = new SolutionBuildCompleteListener();
+            await using var _ = await SubscribeSolutionEventsAsync(solutionBuildCompleteListener);
+            var result = _vsSolutionBuildManager2.StartSimpleUpdateProjectConfiguration(
+                pIVsHierarchyToBuild: projectHierarchy,
+                pIVsHierarchyDependent: null,
+                pszDependentConfigurationCanonicalName: null,
+                dwFlags: (uint)VSSOLNBUILDUPDATEFLAGS.SBF_OPERATION_BUILD,
+                dwDefQueryResults: (uint)VSSOLNBUILDQUERYRESULTS.VSSBQR_SAVEBEFOREBUILD_QUERY_YES,
+                fSuppressUI: 0);
+
+            ErrorHandler.ThrowOnFailure(result);
+
+            return await solutionBuildCompleteListener.WaitForSolutionBuildCompletedAsync(cancellationToken);
+        }
+    }
+
+    private class SolutionBuildCompleteListener() : IVsUpdateSolutionEvents, IDisposable
+    {
+        private TaskCompletionSource<bool> _buildCompletedSource = new();
+
+        public int UpdateSolution_Begin(ref int pfCancelUpdate)
+        {
+            _buildCompletedSource = new TaskCompletionSource<bool>();
```

Comment:
Would be good to validate that any prior `_buildCompletedSource` is correctly completed. Perhaps just call `TrySetCancelled` on it. While that shouldn't be needed, but this code is relying on the caller to do the right thing, which is never ideal for an API. My concern would be that code attempts to wait for a token that's never set.

This sequence could be problematic, for example:

1. Instance of `SolutionBuildCompleteListener` created, which creates `_buildCompletedSource`
1. Code calls `WaitForSolutionBuildCompletedAsync` which waits on `_buildCompletedSource`
1. SBM calls `UpdateSolution_Done` which re-initializes `_buildCompletedSource`
1. Call to `WaitForSolutionBuildCompletedAsync` will never return, leading to a hang

Reply:
If callers to this could potentially be concurrent, consider something like:

```c#
var prior = Interlocked.Exchange(ref _buildCompletedSource, new TaskCompletionSource<bool>());

prior.TrySetCancelled();
```

Reply:
Thanks for catching this, In real world, this situation could happen when user starts another build while this build is not completed.

In this situation, to prevent hanging, maybe instead of creating another `TaskCompletionSource`  in `UpdateSolution_Begin`, we can still use the original `TaskCompletionSource`, and cancel or complete that original `TaskCompletionSource` whenever the original or the second solution build completed/cancelled. The reason why not cancel the original TaskCompletionSource and create a new one is it would cause `WaitForSolutionBuildCompletedAsync` to stop immediately and return false, while in real world, the build might still be succeed.